### PR TITLE
Gen 8 EXP Candies

### PR DESCRIPTION
**Pull request by AgustinGDLV**
Consolidates EXP Candies with Rare Candies, with `holdEffectParam` used to distinguish them.

EXP Candy

## **Discord Contact Info**
Agustin#1522

### CODE REVIEW

**Comment by AsparagusEduardo** [comment]
What happened to all the moves that it would have learned between levels 8 and 24?

**Comment by AgustinGDLV** [comment]
> What happened to all the moves that it would have learned between levels 8 and 24?

Resolved.

File: src/data/items.h [hunk 1]
```diff
@@ -1445,7 +1445,7 @@ const struct Item gItems[] =
         .description = sDynamaxCandyDesc,
         .pocket = POCKET_ITEMS,
         .type = ITEM_USE_BAG_MENU,
-        .fieldUseFunc = ItemUseOutOfBattle_CannotUse, // Todo
+        .fieldUseFunc = ItemUseOutOfBattle_CannotUse,
```
Comment (LOuroboros) [inline src/data/items.h:1448]: Restore this "// Todo". This PR doesn't handle the Dynamax Candy's effect, and it's still not implemented as a result of not having Dynamax mechanics implemented.

File: src/party_menu.c [hunk 2]
```diff
@@ -5067,49 +5075,68 @@ static void DisplayLevelUpStatsPg2(u8 taskId)
...
+        for (; sInitialLevel < sFinalLevel; sInitialLevel++)
+        {
+            SetMonData(&gPlayerParty[gPartyMenu.slotId], MON_DATA_LEVEL, &sInitialLevel);
+            learnMove = MonTryLearningNewMove(&gPlayerParty[gPartyMenu.slotId], TRUE);
+            gPartyMenu.learnMoveState = 1;
+            switch (learnMove)
+            {
+            case 0: // No moves to learn
+                if(sInitialLevel >= sFinalLevel)
+                    PartyMenuTryEvolution(taskId);
```
Comment (LOuroboros) [inline src/party_menu.c:5092]:
```suggestion
                if (sInitialLevel >= sFinalLevel)
                    PartyMenuTryEvolution(taskId);
```

File: src/party_menu.c [hunk 3]
```diff
@@ -5067,49 +5075,68 @@ static void DisplayLevelUpStatsPg2(u8 taskId)
...
+        SetMonData(&gPlayerParty[gPartyMenu.slotId], MON_DATA_LEVEL, &sInitialLevel);
+        result = MonTryLearningNewMove(&gPlayerParty[gPartyMenu.slotId], FALSE);
         gPartyMenu.learnMoveState = 1;
-        switch (learnMove)
+        switch (result)
         {
         case 0: // No moves to learn
-            PartyMenuTryEvolution(taskId);
+            if(sInitialLevel >= sFinalLevel)
+                PartyMenuTryEvolution(taskId);
```
Comment (LOuroboros) [inline src/party_menu.c:5122]:
```suggestion
            if (sInitialLevel >= sFinalLevel)
                PartyMenuTryEvolution(taskId);
```

File: src/party_menu.c [hunk 4]
```diff
@@ -5067,49 +5075,68 @@ static void DisplayLevelUpStatsPg2(u8 taskId)
...
-}
-
-static void Task_TryLearningNextMove(u8 taskId)
-{
-    u16 result = MonTryLearningNewMove(&gPlayerParty[gPartyMenu.slotId], FALSE);
-
-    switch (result)
-    {
-    case 0: // No moves to learn
+    if(sInitialLevel >= sFinalLevel)
```
Comment (LOuroboros) [inline src/party_menu.c:5136]:
```suggestion
    if (sInitialLevel >= sFinalLevel)
```

File: src/pokemon.c [hunk 5]
```diff
@@ -4877,7 +4882,21 @@ bool8 PokemonUseItemEffects(struct Pokemon *mon, u16 item, u8 partyIndex, u8 mov
             if ((itemEffect[i] & ITEM3_LEVEL_UP)
              && GetMonData(mon, MON_DATA_LEVEL, NULL) != MAX_LEVEL)
             {
-                dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
+                if (ItemId_GetHoldEffectParam(item) & FULL_LEVEL)
+                    dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
+                else
```
Comment (LOuroboros) [inline src/pokemon.c:4887]:
```suggestion
                if (ItemId_GetHoldEffectParam(item) & FULL_LEVEL)
                {
                    dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
                }
                else
```

**Review by LOuroboros — CHANGES_REQUESTED** [review]

File: src/strings.c [hunk 6]
```diff
@@ -417,6 +417,7 @@ const u8 gText_PkmnRegainhedHealth[] = _("{STR_VAR_1} regained health.{PAUSE_UNT
 const u8 gText_PkmnBecameHealthy[] = _("{STR_VAR_1} became healthy.{PAUSE_UNTIL_PRESS}");
 const u8 gText_MovesPPIncreased[] = _("{STR_VAR_1}'s PP increased.{PAUSE_UNTIL_PRESS}");
 const u8 gText_PkmnElevatedToLvVar2[] = _("{STR_VAR_1} was elevated to\nLv. {STR_VAR_2}.");
+const u8 gText_PkmnGainedExp[] = _("{STR_VAR_1} gained EXP. POINTS.{PAUSE_UNTIL_PRESS}");
```
Comment (LOuroboros) [inline src/strings.c:420]:
```suggestion
const u8 gText_PkmnGainedExp[] = _("{STR_VAR_1} gained Exp. Points!{PAUSE_UNTIL_PRESS}");
```
https://youtu.be/rsFrXzb-0j8?t=102

File: src/pokemon.c [hunk 7]
```diff
@@ -4877,7 +4882,23 @@ bool8 PokemonUseItemEffects(struct Pokemon *mon, u16 item, u8 partyIndex, u8 mov
             if ((itemEffect[i] & ITEM3_LEVEL_UP)
              && GetMonData(mon, MON_DATA_LEVEL, NULL) != MAX_LEVEL)
             {
-                dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
+                if (ItemId_GetHoldEffectParam(item) & FULL_LEVEL)
+                {
+                    dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
+                }
```
Comment (Sneed69) [inline src/pokemon.c:4888]:
```suggestion
                    SetMonData(mon, MON_DATA_EXP, &dataUnsigned);
                    CalculateMonStats(mon);
                }
```

File: src/pokemon.c [hunk 8]
```diff
@@ -4877,7 +4882,23 @@ bool8 PokemonUseItemEffects(struct Pokemon *mon, u16 item, u8 partyIndex, u8 mov
...
+                else
+                {
+                    temp1 = ItemId_GetHoldEffectParam(item) & 0x0F;
```
Comment (Sneed69) [inline src/pokemon.c:4891]:
```suggestion
                    u32 species = GetMonData(mon, MON_DATA_SPECIES, NULL);

                    temp1 = ItemId_GetHoldEffectParam(item) & 0x0F;
```

File: src/pokemon.c [hunk 9]
```diff
@@ -4877,7 +4882,23 @@ bool8 PokemonUseItemEffects(struct Pokemon *mon, u16 item, u8 partyIndex, u8 mov
...
+                    if (ItemId_GetHoldEffectParam(item) & ONE_THOUSAND)
+                        temp1 *= 1000;
+                    if (ItemId_GetHoldEffectParam(item) & TEN_THOUSAND)
+                        temp1 *= 10000;
+                    dataUnsigned = temp1 + GetMonData(mon, MON_DATA_EXP, NULL);
+                    if (dataUnsigned > gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][MAX_LEVEL])
+                        dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][MAX_LEVEL];
+                }
                 SetMonData(mon, MON_DATA_EXP, &dataUnsigned);
                 CalculateMonStats(mon);
```
Comment (Sneed69) [inline src/pokemon.c:4903]:
```suggestion
                    if (dataUnsigned > gExperienceTables[gBaseStats[species].growthRate][MAX_LEVEL])
                        dataUnsigned = gExperienceTables[gBaseStats[species].growthRate][MAX_LEVEL];
                    SetMonData(mon, MON_DATA_EXP, &dataUnsigned);
                    if (dataUnsigned >= gExperienceTables[gBaseStats[species].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1])
                        CalculateMonStats(mon);
                }
```

**Review by Sneed69 — CHANGES_REQUESTED** [review]

File: src/pokemon.c [hunk 10]
```diff
@@ -4877,9 +4882,30 @@ bool8 PokemonUseItemEffects(struct Pokemon *mon, u16 item, u8 partyIndex, u8 mov
...
+                if (ItemId_GetHoldEffectParam(item) & FULL_LEVEL)
+                {
+                    dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
+                    SetMonData(mon, MON_DATA_EXP, &dataUnsigned);
+                    CalculateMonStats(mon);
+                }
+                else
+                {
+                    u32 species = GetMonData(mon, MON_DATA_SPECIES, NULL);
+
```
Comment (Sneed69) [inline src/pokemon.c:4894]:
```suggestion

                    gBattleScripting.levelUpHP = 0; // Prevents reviving mons that don't level up.
```

**Comment by Jaizu** [comment]
Spherical Ice pointed out that this defines should have some parentheses 
image

File: src/pokemon.c [hunk 11]
```diff
@@ -4885,15 +4885,15 @@ bool8 PokemonUseItemEffects(struct Pokemon *mon, u16 item, u8 partyIndex, u8 mov
                 if (param == 0xFF) // Rare Candy
                 {
                     dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
+                    SetMonData(mon, MON_DATA_EXP, &dataUnsigned);
+                    CalculateMonStats(mon);
```
Comment (AsparagusEduardo) [inline src/pokemon.c:4889]: Why did you move this so it only sets the EXP with Rare Candies and not Exp. candies?

File: src/pokemon.c [hunk 12]
```diff
@@ -4878,24 +4877,18 @@ bool8 PokemonUseItemEffects(struct Pokemon *mon, u16 item, u8 partyIndex, u8 mov
...
              && GetMonData(mon, MON_DATA_LEVEL, NULL) != MAX_LEVEL)
             {
-                if (ItemId_GetHoldEffectParam(item) & FULL_LEVEL)
+                u8 param = ItemId_GetHoldEffectParam(item);
+                if (param == 0xFF) // Rare Candy
                 {
                     dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
                 }
-                else
+                else if (param < NUM_EXP_CANDY_SIZES) // EXP Candies
```
Comment (AsparagusEduardo) [inline src/pokemon.c:4891]: `NUM_EXP_CANDY_SIZES` could be replaced by `ARRAY_COUNT(sExpCandyExperienceTable)`

File: src/pokemon.c [hunk 13]
```diff
@@ -4708,6 +4708,9 @@ bool8 ExecuteTableBasedItemEffect(struct Pokemon *mon, u16 item, u8 partyIndex,
     }                                                                                                   \
 }
 
+// EXP candies store an index for this table in their holdEffectParam.
+static const u32 sExpCandyExperienceTable[] = {100, 800, 3000, 10000, 30000};
```
Comment (AsparagusEduardo) [inline src/pokemon.c:4712]: Maybe we could have this table set up like this:
```suggestion
static const u32 sExpCandyExperienceTable[] = {
    [EXP_100 - 1] = 100,
    [EXP_800 - 1] = 800,
    [EXP_3000 - 1] = 3000,
    [EXP_10000 - 1] = 10000,
    [EXP_30000 - 1] = 30000,
};
```
Then having the defines like this so the default value for Rare Candies is still 0.
```
#define EXP_100 1
#define EXP_800 2
#define EXP_3000 3
#define EXP_10000 4
#define EXP_30000 5
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

File: include/constants/items.h [hunk 14]
```diff
@@ -973,6 +973,13 @@
 #define MACH_BIKE 0
 #define ACRO_BIKE 1
 
+// Item parameters for EXP Candies
+#define EXP_100 1
+#define EXP_800 2
+#define EXP_3000 3
+#define EXP_10000 4
+#define EXP_30000 5
```
Comment (LOuroboros) [inline include/constants/items.h:981]:
```suggestion
#define EXP_100     1
#define EXP_800     2
#define EXP_3000   3
#define EXP_10000 4
#define EXP_30000 5
```

File: src/data/items.h [hunk 15]
```diff
@@ -1359,6 +1359,7 @@ const struct Item gItems[] =
         .name = _("Rare Candy"),
         .itemId = ITEM_RARE_CANDY,
         .price = 10000,
+        .holdEffectParam = 0,
```
Comment (LOuroboros) [inline src/data/items.h:1362]: Is there a point in this? The default value of an item's field is 0 and it remains 0 if it's not given a different value, so... :eyes:

**Review by LOuroboros — CHANGES_REQUESTED** [review]

File: src/data/items.h (hunk 15 above)
Comment (AsparagusEduardo) [inline src/data/items.h:1362]:
```suggestion
```

File: src/pokemon.c [hunk 16]
```diff
@@ -4873,11 +4882,22 @@ bool8 PokemonUseItemEffects(struct Pokemon *mon, u16 item, u8 partyIndex, u8 mov
...
-                dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
+                u8 param = ItemId_GetHoldEffectParam(item);
+                if (param == 0) // Rare Candy
+                {
+                    dataUnsigned = gExperienceTables[gBaseStats[GetMonData(mon, MON_DATA_SPECIES, NULL)].growthRate][GetMonData(mon, MON_DATA_LEVEL, NULL) + 1];
+                }
+                else if (param < ARRAY_COUNT(sExpCandyExperienceTable)) // EXP Candies
+                {
+                    u16 species = GetMonData(mon, MON_DATA_SPECIES, NULL);
+                    dataUnsigned = sExpCandyExperienceTable[param] + GetMonData(mon, MON_DATA_EXP, NULL);
```
Comment (AsparagusEduardo) [inline src/pokemon.c:4897]: I forgot to add this suggestion originally, otherwise it would've shifted the amount of experience.
```suggestion
                    dataUnsigned = sExpCandyExperienceTable[param - 1] + GetMonData(mon, MON_DATA_EXP, NULL);
```

Comment (AgustinGDLV) [reply src/pokemon.c]: Whoops, my bad.

**Review by AsparagusEduardo — APPROVED** [review]
